### PR TITLE
SA-2864: Update CrowdStrike Install Command Scripts to Utilize CrowdStrike API

### PR DESCRIPTION
**Pull request by gweinjc**
## Issues
* [SA-2864](https://jumpcloud.atlassian.net/browse/SA-2864) - Update CrowdStrike Install Command Scripts to Utilize CrowdStrike API

## What does this solve?
Previously, we were instructing customers to download the sensor file and host it publicly in a different location so that the local workstations can access the URL. 

With the JumpCloud x CrowdStrike partnership, customers should be able to utilize CrowdStrike's API to find the latest version of the installer and download it directly from their resources rather than having to host it in an outside location.

## Is there anything particularly tricky?
When testing on Mac devices, ensure that the **CrowdStrike Falcon MDM Policy is correctly applied to devices BEFORE running the command**, otherwise it will not work

## How should this be tested?
1. Navigate to your CrowdStrike console
2. Create a new CrowdStrike API Client with Sensor Download - Read Scope
    * Click left hamburger menu
    * Select **Support and Resources**
    * Under **Resources and Tools**, select **API clients and keys**
    * Select **Add new API Client**
        * Fill out Client Name and Description
        * Scroll down to **Sensor Download** and check the **Read** scope box
        * Select **Add**
3. Save the **Client ID** and **Secret** and **Base URL** Information for future use
4. Create a new Command in the JumpCloud console for each of the installations (Mac and Windows)
    * Make the following adjustments either directly in the Command or in the command settings
        * **MAC** - Run As Root
        * **WINDOWS** - Check Windows PowerShell
        * Replace **CSBaseAddress**, **CSClientID** and **CSClientSecret** with information saved in **Step 3**.
        * Change Timeout Period to something in the neighborhood of 300-600 seconds
5. Assign the command to devices
6. Execute command

## Screenshots
image
image

### CODE REVIEW

File: PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Install CrowdStrike Falcon Agent.md [hunk 1]
```diff
@@ -10,23 +10,71 @@ mac
...
 cd /tmp/$TempFolder
 
-# Download File into Temp Folder
-curl -o "$fileName" "$DownloadUrl"
+# Download the client.
+echo "Beginning CrowdStrike Sensor Download..."
+download=$( /usr/bin/curl -H "Accept: application/octet-stream" -H "Authorization: bearer $bearer" -o "$fileName" "$sensordl?id=$sensorsha")
+res=$?
+
+if test "$res" != "0"; then
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Install CrowdStrike Falcon Agent.md:71]: Does this line work? I'm curious what this does.

We've got brackets on our if statements in line 49 we should use the same syntax for our conditional statements.

Comment (jworkmanjc) [reply PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Install CrowdStrike Falcon Agent.md:71]:
```suggestion
if [[ "$res" != "0" ]]; then
```

would this be correct??

File: PowerShell/JumpCloud Commands Gallery/Windows Commands/Windows - Install CrowdStrike Falcon Agent.md [hunk 2]
```diff
@@ -49,14 +166,15 @@ exit $installerProcess.ExitCode
...
 
 In order to use this command:
 
-1. Download the CrowdStrike Falcon Agent installer and host it at a URL that your devices can access.
-2. Edit the first two lines of the script to include your Customer ID (with checksum value) and the URL where you are hosting the installer.
+1. Create a CrowdStrike API Client with the "SENSOR DOWNLOAD" Read scope and make note of the ClientID and ClientSecret Refer to CrowdStrike's article [Getting Access to the CrowdStrike API](https://www.crowdstrike.com/blog/tech-center/get-access-falcon-apis/) for further information
+2. Set the 3 variables (CSBaseAddress, CSClientID, CSClientSecret) to their respective values for your CrowdStrike API Client
 3. Extend the command timeout to a value that makes sense in your environment. The suggested command timeout for an environment with average network speeds on devices with average computing power is 10 minutes. Note that the command may timeout with a 124 error code in the command result window if not extended, but the script will continue to run.
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Commands Gallery/Windows Commands/Windows - Install CrowdStrike Falcon Agent.md:177]: Yes the first time I ran this I did see the 124 error but the command did complete and the host showed up in my console

File: PowerShell/JumpCloud Commands Gallery/Windows Commands/Windows - Install CrowdStrike Falcon Agent.md [hunk 3]
```diff
@@ -1,20 +1,138 @@
...
+            exit 429
+        }
+        if ($Response.StatusCode -eq 201) {
+            Write-Host "Successfully authenticated; Access Token created"
+            $CrowdStrikeAccessToken = [regex]::Matches($Response.Content, '"(?<name>access_token)": "(?<access_token>.*)",')[0].Groups['access_token'].Value
+            $global:CrowdStrikeAccessToken = $CrowdStrikeAccessToken
+        }
+    }
+}
+function Get-CrowdStrikeCcid {
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Commands Gallery/Windows Commands/Windows - Install CrowdStrike Falcon Agent.md:57]: I really love this, this is nice to not specify this information

File: PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Install CrowdStrike Falcon Agent.md [hunk 4]
```diff
@@ -47,15 +95,15 @@ echo "Deleted /tmp/$TempFolder"
 
 #### Description
 
-This command will download and install the CrowdStrike Falcon Agent to the device if it isn't already installed.
+This command will download and install the CrowdStrike Falcon Agent to the device if it isn't already installed. The command will leverage CrowdStrike's API to find and download the latest version of the Falcon Agent onto the local machine.
 
 In order to use this command, follow the instructions from the [Installing the CrowdStrike Falcon Agent KB](https://support.jumpcloud.com/s/article/Installing-the-Crowdstrike-Falcon-Agent)
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Install CrowdStrike Falcon Agent.md:100]: We should credit the author of this workflow and say we based our bash implementation on Richard Purves' work https://richard-purves.com/2022/05/03/downloading-crowdstrike-via-api-for-fun-and-profit/

File: PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Install CrowdStrike Falcon Agent.md [hunk 5]
```diff
@@ -47,15 +95,15 @@ echo "Deleted /tmp/$TempFolder"
...
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Install CrowdStrike Falcon Agent.md:101]:
```suggestion

[Richard Purves' blog on Downloading Crowdstrike](https://richard-purves.com/2022/05/03/downloading-crowdstrike-via-api-for-fun-and-profit/) inspired this command template.

```

File: PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Install CrowdStrike Falcon Agent.md [hunk 6]
```diff
@@ -10,23 +10,71 @@ mac
 
 ```bash
 #!/bin/bash
-DownloadUrl="https://path/to/url.pkg"
+CSBaseAddress=""
```
Comment (kmaranionjc) [inline PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Install CrowdStrike Falcon Agent.md:13]: Minor thing. Maybe we can put a comment above the variables to be edited that says, `#### Edit these variables`

File: PowerShell/JumpCloud Commands Gallery/Windows Commands/Windows - Install CrowdStrike Falcon Agent.md (hunk 2 above)
Comment (kmaranionjc) [reply PowerShell/JumpCloud Commands Gallery/Windows Commands/Windows - Install CrowdStrike Falcon Agent.md:177]: I got the same exact issue

File: PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Install CrowdStrike Falcon Agent.md (hunk 4 above)
Comment (gweinjc) [reply PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Install CrowdStrike Falcon Agent.md:100]: Resolved with suggestion below!

**Review by jworkmanjc — APPROVED** [review]
Looks good let's get this out a hair before we push the KBs but otherwise tested and approved

**Review by kmaranionjc — APPROVED** [review]

**Pull request merged**